File: Leetcode-Java/BinaryTreePostorderTraversal.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

/**
 * Created by yhf on 15/6/12.
 */

/*
Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [3,2,1].

Note: Recursive solution is trivial, could you do it iteratively?
*/

public class BinaryTreePostorderTraversal {

    public static List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> cur = new ArrayList<>();
        postorderTraversal(root, cur);
        return cur;
    }

    private static void postorderTraversal(TreeNode root, List<Integer> cur) {
        // Recursive
        if (root != null) {
            postorderTraversal(root.left, cur);
            postorderTraversal(root.right, cur);
            cur.add(root.val);
        }
    }

    public static List<Integer> postorderTraversal2(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        List<Integer> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        stack.push(root);
        TreeNode prev = null;

        while (!stack.isEmpty()) {
            TreeNode cur = stack.peek();
            if (prev == null || prev.left == cur || prev.right == cur) {
                // go down the tree
                if (cur.left != null) {
                    stack.push(cur.left);
                } else if (cur.right != null) {
                    stack.push(cur.right);
                } else {
                    res.add(stack.pop().val);
                }
            } else if (prev == cur.left) {
                // go up from left child
                if (cur.right != null) {
                    stack.push(cur.right);
                } else {
                    res.add(stack.pop().val);
                }
            } else {
                // go up from right child
                res.add(stack.pop().val);
            }
            prev = cur;
        }

        return res;
    }

    public static void main(String[] args) {
        TreeNode n1 = new TreeNode(1);
        TreeNode n2 = new TreeNode(2);
        TreeNode n3 = new TreeNode(3);
        n1.right = n2;
        n2.left = n3;
        List<Integer> list = postorderTraversal2(n1);
    }
}

File: Leetcode-Python/BinaryTreePostorderTraversal.py
"""
Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [3,2,1].

Note: Recursive solution is trivial, could you do it iteratively?
"""
from TreeNode import TreeNode


class Solution:
    # @param {TreeNode} root
    # @return {integer[]}
    def postorderTraversal(self, root):

        self.__traversal = []

        def __postorderTraversal(root):
            if not root:
                return
            __postorderTraversal(root.left)
            __postorderTraversal(root.right)
            self.__traversal.append(root.val)

        __postorderTraversal(root)
        return self.__traversal

    def postorderTraversalBFS(self, root):
        res = []
        if not root:
            return res
        stack = [root]
        prev = None
        while stack:
            cur = stack[-1]
            if not prev or prev.left == cur or prev.right == cur:  # going down
                if cur.left:
                    stack.append(cur.left)
                elif cur.right:
                    stack.append(cur.right)
                else:
                    res.append(stack.pop().val)
            elif prev == cur.left:                                 # going up from left child
                if cur.right:
                    stack.append(cur.right)
                else:
                    res.append(stack.pop().val)
            else:                                                  # going up from right child
                res.append(stack.pop().val)
            prev = cur

        return res

if __name__ == '__main__':
    n1 = TreeNode(1)
    n2 = TreeNode(2)
    n3 = TreeNode(3)
    n4 = TreeNode(4)
    n5 = TreeNode(5)
    n6 = TreeNode(6)
    n7 = TreeNode(7)
    n8 = TreeNode(8)
    n1.left = n2
    n1.right = n3
    n2.left = n4
    n2.right = n5
    n3.left = n6
    n3.right = n7
    n4.left = n8

    s = Solution()
    assert s.postorderTraversal(n1) == [8, 4, 5, 2, 6, 7, 3, 1]